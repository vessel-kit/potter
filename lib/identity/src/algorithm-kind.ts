import { InvalidAlgorithmKindError } from "./invalid-algorithm-kind.error";

/**
 * Supported [JOSE algorithms](https://www.iana.org/assignments/jose/jose.xhtml#web-signature-encryption-algorithms).
 */
export enum AlgorithmKind {
  /**
   * `secp256k1` signing of `SHA-256` digest as specified in [RFC 8812](https://tools.ietf.org/html/rfc8812).
   */
  ES256K = "ES256K",
  /**
   * `ed25519` signing as specified in [RFC 8037](https://tools.ietf.org/html/rfc8037).
   */
  EdDSA = "EdDSA",
}

/* istanbul ignore next */
export namespace AlgorithmKind {
  /**
   * Convert unknown external `input` to [[AlgorithmKind]].
   */
  export function fromString(input: string): AlgorithmKind {
    const searching = input as AlgorithmKind;
    switch (searching) {
      case AlgorithmKind.EdDSA:
        return AlgorithmKind.EdDSA;
      case AlgorithmKind.ES256K:
        return AlgorithmKind.ES256K;
      default:
        throw new InvalidAlgorithmKindError(searching);
    }
  }
}
