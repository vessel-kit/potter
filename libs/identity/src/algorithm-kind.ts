import { InvalidAlgorithmKindError } from './invalid-algorithm-kind.error';

export enum AlgorithmKind {
  ES256K = 'ES256K',
  Ed25519 = 'Ed25519',
}

/* istanbul ignore next */
export namespace AlgorithmKind {
  export function fromString(input: string) {
    const searching = input as AlgorithmKind;
    switch (searching) {
      case AlgorithmKind.Ed25519:
        return AlgorithmKind.Ed25519;
      case AlgorithmKind.ES256K:
        return AlgorithmKind.ES256K;
      default:
        throw new InvalidAlgorithmKindError(searching);
    }
  }
}
