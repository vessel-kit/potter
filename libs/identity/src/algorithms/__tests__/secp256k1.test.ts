import * as secp256k1 from '../secp256k1';
import * as _ from 'lodash';
import { AlgorithmKind } from '../../algorithm-kind';

describe('PublicKey', () => {
  test('fields', () => {
    const material = new Uint8Array(_.times(32, () => 1));
    const key = new secp256k1.PublicKey(material);
    expect(key.kind).toEqual(AlgorithmKind.secp256k1);
    expect(key.material).toEqual(material);
  });
});

describe('PrivateKey', () => {
  const material = new Uint8Array(_.times(32, () => 1));
  const key = new secp256k1.PrivateKey(material);
  test('fields', async () => {
    expect(key.kind).toEqual(AlgorithmKind.secp256k1);
    const publicKey = await key.publicKey();
    expect(publicKey).toMatchSnapshot();
  });
  test('sign', async () => {
    const signature = await key.sign(material);
    expect(signature).toMatchSnapshot();
  });
});

describe('verifySignature', () => {
  const material = new Uint8Array(_.times(32, () => 1));
  const key = new secp256k1.PrivateKey(material);
  const message = new Uint8Array(_.times(32, (n) => n));
  test('ok', async () => {
    const publicKey = await key.publicKey();
    const signature = await key.sign(message);
    expect(secp256k1.verifySignature(publicKey, message, signature)).toBeTruthy();
  });
  test('wrong input', async () => {
    const publicKey = await key.publicKey();
    const signature = await key.sign(message);
    expect(secp256k1.verifySignature(publicKey, new Uint8Array(), signature)).toBeFalsy();
  });
});
