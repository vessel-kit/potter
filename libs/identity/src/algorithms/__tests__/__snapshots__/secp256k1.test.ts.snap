// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PrivateKey fields 1`] = `
PublicKey {
  "kind": "secp256k1",
  "material": Uint8Array [
    3,
    27,
    132,
    197,
    86,
    123,
    18,
    100,
    64,
    153,
    93,
    62,
    213,
    170,
    186,
    5,
    101,
    215,
    30,
    24,
    52,
    96,
    72,
    25,
    255,
    156,
    23,
    245,
    233,
    213,
    221,
    7,
    143,
  ],
}
`;

exports[`PrivateKey sign 1`] = `
Uint8Array [
  198,
  75,
  25,
  36,
  21,
  119,
  72,
  101,
  39,
  51,
  196,
  18,
  148,
  229,
  246,
  227,
  149,
  195,
  98,
  106,
  142,
  145,
  31,
  55,
  66,
  164,
  184,
  173,
  79,
  219,
  146,
  47,
  52,
  123,
  165,
  205,
  214,
  41,
  2,
  126,
  245,
  132,
  110,
  184,
  92,
  148,
  82,
  246,
  49,
  46,
  10,
  234,
  105,
  118,
  37,
  214,
  107,
  32,
  36,
  72,
  243,
  233,
  97,
  143,
]
`;
